### ISSUE TITLE
Allow simple expressions in SetEnvironmentVariables/SetSystemProperties

### ISSUE THREAD

**Issue by josiahhaswell** [junit-pioneer/junit-pioneer#450]
Dynamically computing environment variable values would be very useful for a wide variety of tests.   For instance:

```
@SetEnvironmentVariable(key = "MY_ENV", value = "${TestUtilities.resolveClassPathFilePath('test.xml')}")
void testTestXml() {
// etc. 
}
```

would allow the SetEnvironmentVariables/SetSystemProperties annotations to dynamically resolve environment variable values from the runtime context.

An alternative might be to allow MVEL/Groovy/SPEL (or whatever) expression evaluation in for the value parameter of the annotation.  

I'd be happy to contribute a PR if there's a direction you'd like to go with this

**Comment by beatngu13** [comment]
Hey @josiahhaswell,

thx for reaching out!

Dynamically computed values are already possible, at least partly. As stated in the [system property docs](https://junit-pioneer.org/docs/system-properties/):

>Sometimes, you might also need to set a system property to a value that is not a constant expression, which is required for annotation values. In this case, you can still leverage the restore mechanism:
>
> ```java
> @ParameterizedTest
> @ValueSource(strings = { "foo", "bar" })
> @ClearSystemProperty(key = "some property")
> void test(String value) {
> 	System.setProperty("some property", value);
> }
> ```

However, this is not possible for environment variables, yet – see #381.

I see two things to discuss:

1. Using `@ClearEnvironmentVariable`/`@ClearSystemProperty` for dynamic values / non-constant expressions might not be obvious, so we could think about an additional annotation. For instance, something like `@RestoreEnvironmentVariable`/`@RestoreSystemProperty`, similar to [System Rules](https://stefanbirkner.github.io/system-rules/#RestoreSystemProperties). (They would do the same as their `@Clear*` counterparts, they are just there for readability purposes.)
2. Whether or not we want to support expressions. Personally, I'm not sure about this. Having method references just like JUnit does for e.g. `@MethodSource` is one thing. But full support for expressions may quickly grow in terms of complexity, dependencies, … anyway, this is just what I think. Let's leave this up for this discussion.

**Comment by josiahhaswell** [comment]
So, we implemented a prototype using Spring MVEL (I'm definitely not recommending bringing that in as a dependency here) that works well.  MVEL, which is self-contained, might be a feasible option in terms of reducing dependencies.  Each of these solutions suffers from the complexity of defining an execution context for the expression language, although SPEL automatically provides a sensible one, viz. the ApplicationContext.


To be honest, in the course of prototyping this in our own testing infrastructure, we found that a method-call accepting a Consumer of an environment abstraction worked well.  We avoided concurrency issues by completely swapping out the underlying map(s) in the environment for the duration of an invocation a la:
```

  @Test
  void ensurePropertiesApiMakesSense() {
    val key =
        secureRandomString(40); // avoid collisions with users' environments
    EnvironmentManager.props()
        .key(key)
        .value("world")
        .key(key + 4)
        .value("sup")
        .with(
            e -> {
              assertEquals(e.getSystemProperty(key), "world");
              assertEquals(e.getSystemProperty(key + 4), "sup");
            });
    assertNull(Environment.getDefault().getSystemProperty(key));
    assertNull(Environment.getDefault().getSystemProperty(key + 4));
  }

```

While this _supports_ SPEL expressions, we're discovering that they aren't very useful in this context, so a `MethodSource` based solution would almost certainly be fine.
